_drafts/2024-05-18-get-started-with-aws-tf-01.md

```diff
@@ -0,0 +1,254 @@
+
+Now try running `terraform apply` once more and everything should be hunky dory.
+
+## Step 3 - Connect to our EC2 Instance
+
+If you've made it this far it means you've successfully configured terraform with an aws account and have deployed an EC2 instance to AWS. Nice!  
+Let's try connect to this instance.  
+We can find the ip address by logging in to the AWS Management Console and opening the instance details within EC2.
+Copy the IPv4 address and in a terminal run `ping <the-ip-address>`.
+
+No response...  
+
+This is because when we define the EC2 instance, a default security group is applied to it. This security group allows incoming connections on any port using any protocol but only from other aws resources with the same security group rules...
+
+If we want to connect to our instance, we need to make our own rules.
+
+When defining a security group rule, we declare if the rule is an ingress or egress rule (incoming or outgoing). We can also declare which port the rule applies to and the ip address range.
+
+We don't want just anyone to connect to this service so for now we'll only allow our local device's public ip address in the address range. You can find this easily using a site like [whatismyipaddress.com](www.whatismyipaddress.com). 
+
+Open `unicon-api.tf` and add the following resources, providing your public ip for `cidr_ipv4` value:
+
+```terraform
+# /infra/unicorn-api.tf
+
+resource "aws_security_group" "unicorn_api_security_group" {
+  name = "unicorn-api-security-group"
+}
+
+resource "aws_vpc_security_group_ingress_rule" "allow_from_local_device" {
+  security_group_id = aws_security_group.unicorn_api_security_group.id
+  description       = "allow ingress on any port from any ip"
+  cidr_ipv4         = "<YOUR_IP_ADDRESS_HERE>/32"
+  ip_protocol       = "tcp"
+  from_port         = 22
+  to_port           = 22
+}
+```
+
+Next, apply the security group to the ec2 instance, adding the following property to the ec2 instance:
+
+```terraform 
+# /infra/unicorn-api.tf
+
+resource "aws_instance" "web_service_vm" {
+
+    # ...
+
+  vpc_security_group_ids = [aws_security_group.unicorn_api_security_group.id]
+}
+```
+
+With these new resources defined, we can re-apply our changes with `terraform apply`. This will rebuild the EC2 instance, so we'll need to get the latest ip address so we can ping it.  
+Once the resources have been updated and the latest IP address copied, we can now ping our EC2 instance with:  
+
+`ping <the-ip-address>`
+
+## Step 4 - SSH into the EC2 Instance
+
+In this step we're going to create an ssh key pair using terraform. The public key will be added to the ec2 instance as one of its known ssh public keys.  
+We'll save the private key as a local file so we can reference it when we try connect to the ec2 instance.
+
+First, add the following providers to our `main.tf`:
```

Comment:
It'd be worth offering some basic context around why an extra provider is needed here and how the `tls` provider differs from the `aws` provider - namely the `aws` provider largely manages resources "in the cloud" and the `tls` provider provides utilities for managing TLS keys/certificates.

[TLS Provider Docs](https://registry.terraform.io/providers/hashicorp/tls/latest/docs)